### ISSUE TITLE
MSBuild not found when running from Git Bash

### ISSUE THREAD

**Issue by briandonahue** [Albacore/albacore#156]
At first I thought #140 would fix this, but I think my issue is unrelated.  MSBuild 12 runs fine in a windows cmd shell, but I'd like to run in Git Bash if possible.  It seems that the "where" call here fails, "No such file or directory." I tried adding C:\Windows\System32 to my .bashrc path, that didn't work.  It *does* work if I hard code the full path on line 192 of cross_platform_cmd.rb:

      cmd = ::Rake::Win32.windows? ? 'c:\windows\system32\where' : 'which'

I'm fully aware this could be a case of my shell ignorance, or simply "don't use git bash" but wondering if anyone could shed any light on a better solution.  Happy to create a pull request and/or do some digging if anyone has some ideas?

**Comment by haf** [comment]
I always used git bash when I was using windows, with this code in Albacore, without a problem, though.

**Comment by briandonahue** [comment]
I have seen similar issues calling out to commands, but I thought I always fixed it by adding the path to .bashrc, but I did that, restarted the shell, and still had the same issue. :confused:

**Comment by AnthonyMastrean** [comment]
The entire `system32` directory should be in your machine PATH by default. It is in mine and I haven't messed with any Git Bash settings (in `.bashrc` or elsewhere). Can you double check that your PATH contains `system32`?

Here's mine

```bash
$ where where
c:\Windows\System32\where.exe

$ which where
/c/windows/system32/where
```

**Comment by briandonahue** [comment]
@AnthonyMastrean Thanks for replying.  I can run the commend from bash shell, but when I run rake/albacore the command is not found.

    $ where where
    INFO: Could not find files for the given pattern(s).

The above doesn't seem to be a valid pattern for the where command.  This works:

    $ where "c:\windows\system32:where"
    c:\Windows\System32\where.exe

And this works:

    $ which where
    /c/Windows/system32/where

**Comment by briandonahue** [comment]
This also works, from Git Bash:

    $ where "c:\program files (x86)\msbuild\12.0\bin:msbuild.exe"
    c:\program files (x86)\MSBuild\12.0\Bin\MSBuild.exe

**Comment by AnthonyMastrean** [comment]
Oh!

**Comment by haf** [comment]
Are you sure it's not found? It searches more than one location by default, resulting in those messages.

It's only if you get a 127 return code that you can see this causes an error.

**Comment by briandonahue** [comment]
What it's not finding is the `where` command, as though it's not in the path that is running in the script context.  The result of a `where` when the file is not found is different.  I see that result when I run the rake script at a Windows CMD Prompt.  This is specific to when I run rake from Git Bash.

**Comment by briandonahue** [comment]
Also note in the initial description, if I change the line in CrossPlatformCmd to use the full path of `where` it succeeds when running in Bash.  So it seems it has something to do with the path info that is available to albacore/rake in the Bash context.

**Comment by haf** [comment]
It might have something to do with the bitness of your installation of ruby.

**Comment by briandonahue** [comment]
Yeah, this definitely seems to be a "broken on my machine" issue.  Have seen it work in Git Bash on two other machines now.

**Comment by Cyberlane** [comment]
I have this issue when running the latest version, however if I roll back to version 1.0.0 - it works just fine. Can somebody confirm that this still works on the latest version?

**Comment by haf** [comment]
@briandonahue confirmed just above your message.

**Comment by haf** [comment]
Could you try the commands in the IRB?

Here's another run https://ci.appveyor.com/project/AdemarGonzalez/suave/build/1.0.357#L273 that runs on Windows

**Comment by haf** [comment]
@Cyberlane btw, if you want a bit more 'hands on' from me, we can probably talk a few hours worth of consulting; e-mail henrik@haf.se

**Comment by haf** [comment]
Hey, I know; could it have something to do with https://github.com/Albacore/albacore/pull/140 /cc @devployment

**Comment by haf** [comment]
What if we default to the old behaviour if it is not found under the registry key? Could someone try it out and implement it?

**Comment by Cyberlane** [comment]
That might be and yes that might work - however I have just had to roll back to version 0.3.5 as xunit did not work in 1.0.0 either. Everything appears stable in 0.3.5 so I may just leave it be until I have more time to look into the issue and perhaps issue a PR with some fixes.

**Comment by devployment** [comment]
Just scanned the discussion here. But for what I can see it's all related to the 

`cmd = ::Rake::Win32.windows? ? 'where' : 'which'` 

and the fact that the executable is not found as it should be. Correct me if I'm wrong.

As @briandonahue wrote the `where where` gives no executable. I will check but I don't see a relation to my changes. But never say never.

@briandonahue Have you installed git and overwrote the windows tools maybe (as shown in the 3rd option) so that it cannot find `where` anymore? Maybe altered other things related to this?

![](http://s4.postimg.org/6natqumbx/Git_Setup_Run_from_Windows_Command_Prompt.png)

**Comment by briandonahue** [comment]
@devployment Your summary of the issue is correct.  Doesn't find the where executable, in a Git Bash shell.  What's odd is that it works fine in a CMD window.  So, Path is OK there.  

You got me excited for a second, because the middle option was circled, which is what I usually use.  I don't think I've ever done the third option, which I think is what you are saying could break it.

But, it does work from Git Bash on some other machines, where I can't remember what options I used.  Could it be that only the FIRST option `Use Git Bash Only` works?

Here is my path, btw:

````
C:\Program Files (x86)\Parallels\Parallels Tools\Applications;
C:\ProgramData\Oracle\Java\javapath;
C:\tools\ruby193\bin;
C:\Windows\system32;
C:\Windows;
C:\Windows\System32\Wbem;
C:\Windows\System32\WindowsPowerShell\v1.0\;
C:\Program Files (x86)\Microsoft SQL Server\100\Tools\Binn\;
C:\Program Files\Microsoft SQL Server\100\Tools\Binn\;
C:\Program Files\Microsoft SQL Server\100\DTS\Binn\;
C:\Program Files (x86)\Microsoft SQL Server\100\Tools\Binn\VSShell\Common7\IDE\;
C:\Program Files (x86)\Microsoft Visual Studio 9.0\Common7\IDE\PrivateAssemblies\;
C:\Program Files (x86)\Microsoft SQL Server\100\DTS\Binn\;
C:\ProgramData\chocolatey\bin;
C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;
C:\Program Files (x86)\vim\bin;
C:\Program Files\SourceGear\Common\DiffMerge\;
c:\tools\mongodb\bin;
C:\Program Files (x86)\Windows Kits\8.1\Windows Performance Toolkit\;
C:\Program Files (x86)\Microsoft SDKs\TypeScript\1.0\;
C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;
C:\Program Files (x86)\Common Files\Intuit\QBPOSSDKRuntime;
C:\Program Files (x86)\Git\bin\
````

**Comment by briandonahue** [comment]
Still so confused/frustrated by this.  The only difference I can find between an environment where this works and doesn't work is a minor bump in ruby version.  

It works on a box with `ruby 1.9.3p194 (2012-04-20) [i386-mingw32]` but not on my local box with `ruby 1.9.3p545 (2014-02-24) [i386-mingw32]`

Not to mention it works from cmd.exe but not Git Bash (locally, works both on remote box).  To make things more frustrating, this works in irb on my local box from Git Bash:

````
$ irb
irb(main):001:0> `where`
=> "The syntax of this command is:\n\nWHERE [/R dir] [/Q] [/F] [/T] pattern...\n\nDescription:\n    Displays the location of files that match the search pattern.\n    By default, the search is done along the current directory and\n    in the paths specified by the PATH environment variable.\n\nParameter List:\n    /R
  
...

````

So it obviously *can* find it from Ruby.  Next I'll try a simpler rake file with one task and see if it works there.

**Comment by briandonahue** [comment]
OK, so this is something funky in my ancient Rakefile and something in this environment.  I just created a Rakefile with a single msbuild task, and `where` ran as it should from Git Bash...

**Comment by briandonahue** [comment]
This had something to do with my `.bashrc` and/or `.bash_profile` which was also ancient.  The weird thing is that the same files exist on the other box, but it's all outdated path manipulation stuff.  My guess is that something I added to the path didn't exist and was messing things up... not sure.  Anyway, I just deleted them as I didn't need anything right now in them, and `where` now works as it should.
